EIPS/eip-6224.md

```diff
@@ -0,0 +1,244 @@
+    *  @param contractAddress the address of the implementation
+    */
+   function addProxyContract(string memory name, address contractAddress) external;
+
+   /**
+    *  @notice RECOMMENDED The function that adds an already deployed proxy to the ContractsRegistry. It MAY be used
+    *  when the system migrates to the new ContractRegistry. In that case, the new ProxyUpgrader MUST have the
+    *  credentials to upgrade the newly added proxies
+    *  @param name the name to associate the contract with
+    *  @param contractAddress the address of the proxy
+    */
+   function justAddProxyContract(string memory name, address contractAddress) external;
+
+   /**
+    *  @notice REQUIRED The function to remove contracts from the ContractsRegistry
+    *  @param name the associated name with the contract
+    */
+   function removeContract(string memory name) external;
+}
+```
+
+- The `ContractsRegistry` SHOULD deploy the `ProxyUpgrader` contract in the constructor that MUST be set as an admin of `Transparent` proxies deployed via `addProxyContract` method.
+- It SHOULD NOT be possible to add zero addresses to the `ContractsRegistry`.
+- The `ContractsRegistry` MUST use the `IDependant` interface in the `injectDependencies` method.
+
+### Dependant
+
+The `Dependant` contract is the one that depends on other contracts present in the system. In order to support dependency injection mechanism, the dependant contract MUST implement the following interface:
+
+```solidity
+pragma solidity ^0.8.0;
+
+interface IDependant {
+   /**
+    *  @notice The function that is called from the ContractsRegistry (or factory) to inject dependencies.
+    *  @param contractsRegistry the registry to pull dependencies from
+    *
+    *  The Dependant MUST perform a dependency injector access check to this method
+    */
+   function setDependencies(address contractsRegistry) external;
+
+   /**
+    *  @notice The function that sets the new dependency injector.
+    *  @param injector the new dependency injector
+    *
+    *  The Dependant MUST perform a dependency injector access check to this method
+    */
+   function setInjector(address injector) external;
+
+   /**
+    *  @notice The function that gets the current dependency injector
+    *  @return the current dependency injector
+    */
+   function getInjector() external view returns (address);
+}
+```
+
+- The `Dependant` contract MUST pull its dependencies in the `setDependencies` method from the passed `contractsRegistry` address.
+- The `Dependant` contract MAY store the dependency injector address in the special slot `0x3d1f25f1ac447e55e7fec744471c4dab1c6a2b6ffb897825f9ea3d2e8c9be583` (obtained as `bytes32(uint256(keccak256("eip6224.dependant.slot")) - 1)`).
+
+
+## Rationale
+
+The EIP proposes a way to manage the system's smart contracts and maintain a clear usage of their dependencies. There are a few design decisions that have to be specified explicitly:
```

Comment:
```suggestion
```
This is more motivation, and it is assumed that the readers of an EIP know what all the sections mean.